src/serac/physics/contact/contact_data.cpp

```diff
@@ -0,0 +1,294 @@
+  auto         dof_offsets = pressureTrueDofOffsets();
+  for (size_t i{0}; i < pairs_.size(); ++i) {
+    if (pairs_[i].getContactOptions().enforcement == ContactEnforcement::LagrangeMultiplier) {
+      mfem::Vector p_pair_true;
+      p_pair_true.MakeRef(p_true, dof_offsets[static_cast<int>(i)],
+                          dof_offsets[static_cast<int>(i) + 1] - dof_offsets[static_cast<int>(i)]);
+      pairs_[i].pressure().ParFESpace()->GetRestrictionMatrix()->Mult(pairs_[i].pressure(), p_pair_true);
+    }
+  }
+  return p_true;
+}
+
+mfem::Vector ContactData::trueGaps() const
+{
+  mfem::Vector g_true(numPressureTrueDofs());
+  auto         dof_offsets = pressureTrueDofOffsets();
+  for (size_t i{0}; i < pairs_.size(); ++i) {
+    if (pairs_[i].getContactOptions().enforcement == ContactEnforcement::LagrangeMultiplier) {
+      mfem::Vector g_pair_true;
+      g_pair_true.MakeRef(g_true, dof_offsets[static_cast<int>(i)],
+                          dof_offsets[static_cast<int>(i) + 1] - dof_offsets[static_cast<int>(i)]);
+      // NOTE: gaps are considered a dual field in MFEM and the correct operator
+      // here for MFEM dual fields is P^T.  However, Tribol stores gaps similar
+      // to a ParGridFunction -- all shared dofs are equal; therefore, the
+      // operator that returns the desired values is R
+      pairs_[i].pressure().ParFESpace()->GetRestrictionMatrix()->Mult(pairs_[i].gaps(), g_pair_true);
+    }
+  }
+  return g_true;
+}
+
+std::unique_ptr<mfem::BlockOperator> ContactData::contactJacobian() const
+{
+  jacobian_offsets_    = mfem::Array<int>({0, reference_nodes_->ParFESpace()->GetTrueVSize(),
+                                        numPressureTrueDofs() + reference_nodes_->ParFESpace()->GetTrueVSize()});
+  auto block_J         = std::make_unique<mfem::BlockOperator>(jacobian_offsets_);
+  block_J->owns_blocks = true;
+  mfem::Array2D<mfem::HypreParMatrix*> constraint_matrices(static_cast<int>(pairs_.size()), 1);
+
+  for (size_t i{0}; i < pairs_.size(); ++i) {
+    auto pair_J         = tribol::getMfemBlockJacobian(pairs_[i].getPairId());
+    pair_J->owns_blocks = false;
+    if (!pair_J->IsZeroBlock(0, 0)) {
+      SLIC_ERROR_ROOT_IF(!dynamic_cast<mfem::HypreParMatrix*>(&pair_J->GetBlock(0, 0)),
+                         "Only HypreParMatrix constraint matrix blocks are currently supported.");
+      if (block_J->IsZeroBlock(0, 0)) {
+        block_J->SetBlock(0, 0, &pair_J->GetBlock(0, 0));
+      } else {
+        if (block_J->IsZeroBlock(0, 0)) {
+          block_J->SetBlock(0, 0, &pair_J->GetBlock(0, 0));
+        } else {
+          block_J->SetBlock(0, 0,
+                            mfem::Add(1.0, static_cast<mfem::HypreParMatrix&>(block_J->GetBlock(0, 0)), 1.0,
+                                      static_cast<mfem::HypreParMatrix&>(pair_J->GetBlock(0, 0))));
+        }
+        delete &pair_J->GetBlock(0, 0);
+      }
+    }
+    if (!pair_J->IsZeroBlock(1, 0)) {
```

Comment:
I have to admit this is fairly hard to read. Can you add some comments to help walk through what you're doing?